### PR TITLE
TDL-16648: Implement request timeouts

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-16648: Implement request timeouts
- Added request timeout for the tap, default timeout is 300 seconds.

# QA steps
 - [ ] automated tests passing
 - [ ] manual qa steps passing (list below)
 
# Risks

# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_shopify/streams/base.py [hunk 1]
```diff
@@ -91,6 +113,13 @@ class Stream():
     def __init__(self):
         self.results_per_page = Context.get_results_per_page(RESULTS_PER_PAGE)
 
+        self.request_timeout = REQUEST_TIMEOUT # set default timeout
+        timeout_from_config = Context.config.get('request_timeout')
+        # updated the timeout value if timeout is passed in config and not from 0, "0", ""
+        if timeout_from_config and float(timeout_from_config):
```
Comment (dbshah1212) [inline tap_shopify/streams/base.py:119]: Why do we need to write this code 2 times, can't we put this inside one function and use it?

File: tests/unittests/test_timeout.py [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+            "results_per_page": 50,
+            "request_timeout": "100"
+        }
+
+        # initialize base class
+        stream = Stream()
+        # verify the timeout is set as expected
+        self.assertEquals(stream.request_timeout, 100)
+
+    def test_timeout_empty_value_passed_in_config(self):
```
Comment (dbshah1212) [inline tests/unittests/test_timeout.py:67]: Please add function comments

File: tap_shopify/streams/base.py (hunk 1 above)
Comment (harshpatel4crest) [reply tap_shopify/streams/base.py]: Added a function that returns the timeout value.

File: tests/unittests/test_timeout.py (hunk 2 above)
Comment (harshpatel4crest) [reply tests/unittests/test_timeout.py:76]: Added

**Review by dbshah1212 — APPROVED** [review]

File: tap_shopify/streams/base.py [hunk 3]
```diff
@@ -50,7 +65,26 @@ def retry_after_wait_gen(**kwargs):
     sleep_time_str = resp.headers.get('Retry-After', resp.headers.get('retry-after'))
     yield math.floor(float(sleep_time_str))
 
+# boolean function to check if the error is 'timeout' error or not
+def is_timeout_error():
+    """
+        This function checks whether the error contains 'timed out' substring and return boolean
+        values accordingly, to decide whether to backoff or not.
+    """
+    def gen_fn(exc):
```
Comment (dbshah1212) [inline tap_shopify/streams/base.py:74]: do we need inner function? OR one function is enough

File: tap_shopify/streams/inventory_levels.py [hunk 4]
```diff
@@ -10,9 +10,12 @@ class InventoryLevels(Stream):
     replication_key = 'updated_at'
     key_properties = ['location_id', 'inventory_item_id']
     replication_object = shopify.InventoryLevel
+    # Added decorator over functions of shopify SDK
+    replication_object.find = shopify_error_handling(replication_object.find)
 
-    @shopify_error_handling
```
Comment (dbshah1212) [inline tap_shopify/streams/inventory_levels.py:14]: Why do we remove this, isn't it required for other error handling?

File: tap_shopify/streams/locations.py [hunk 5]
```diff
@@ -6,9 +6,12 @@
 class Locations(Stream):
     name = 'locations'
     replication_object = shopify.Location
+    # Added decorator over functions of shopify SDK
+    replication_object.find = shopify_error_handling(replication_object.find)
 
-    @shopify_error_handling
```
Comment (dbshah1212) [inline tap_shopify/streams/locations.py:10]: Why do we remove this, isn't it required for other error handling?

File: tap_shopify/streams/base.py (hunk 3 above)
Comment (harshpatel4crest) [reply tap_shopify/streams/base.py]: Used 1 function for **giveup** condition.

File: tap_shopify/streams/inventory_levels.py (hunk 4 above)
Comment (harshpatel4crest) [reply tap_shopify/streams/inventory_levels.py:14]: As we have added the decorator over the `find` method. Hence removed this decorator.

**Review by dbshah1212 — APPROVED** [review]

File: tap_shopify/streams/base.py [hunk 6]
```diff
@@ -50,7 +65,23 @@ def retry_after_wait_gen(**kwargs):
...
+    # retry if the error string contains 'timed out'
+    if str(error_raised).__contains__('timed out'):
+        return False
+    return True
+
 def shopify_error_handling(fnc):
+    @backoff.on_exception(backoff.expo, # timeout error raise by Shopify
+                          (pyactiveresource.connection.Error, socket.timeout),
+                          giveup=is_timeout_error,
+                          max_tries=MAX_RETRIES,
```
Comment (karanpanchal-crest) [inline tap_shopify/streams/base.py:83]: @harshpatel4crest Don't we have to use `retry_after_wait_gen` function to wait for the next retry?

**Review by karanpanchal-crest — CHANGES_REQUESTED** [review]

File: tap_shopify/streams/base.py (hunk 6 above)
Comment (harshpatel4crest) [reply tap_shopify/streams/base.py:83]: In the code, `retry_after_wait_gen` is used for the request error that contains the header in the error response.

Comment (karanpanchal-crest) [reply tap_shopify/streams/base.py:83]: Ok, so `retry_after_wait_gen` will be used in case of Errors like 429 (Max retries). Got it, Thank you

**Review by karanpanchal-crest — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

File: .circleci/config.yml [hunk 7]
```diff
@@ -25,6 +25,18 @@ jobs:
...
+          command: |
+            source /usr/local/share/virtualenvs/tap-shopify/bin/activate
+            pip install nose coverage
+            nosetests --with-coverage --cover-erase --cover-package=tap_shopify --cover-html-dir=htmlcov tests/unittests
+            coverage html
+      - store_test_results:
+          path: test_output/report.xml
+      - store_artifacts:
+          path: htmlcov
+      - add_ssh_keys
```
Comment (kspeer825) [inline .circleci/config.yml:39]: I don't think this add_ssh_keys line is necessary

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**